[Extensions] Add AOT compatibility attributes to Microsoft.Extensions.Azure

eerhardt opened the pull request:
Annotate the Microsoft.Extensions.Azure library for trimming/AOT compatibility.

Also add a few suppressions to Azure.Identity's EventSource that showed up when I was using Microsoft.Extensions.Azure in a real-world app.

Add CI checks for both `Microsoft.Extensions.Azure` and `Azure.Identity` for AOT compatibility.

Continues work started in #35572.

cc @m-redding

eerhardt committed:
[Extensions] Add AOT compatibility attributes

Annotate the Microsoft.Extensions.Azure library for trimming/AOT compatibility.

Also add a few suppressions to Azure.Identity's EventSource that showed up when I was using Microsoft.Extensions.Azure in a real-world app.

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Microsoft.Extensions.Azure](https://apiview.dev/Assemblies/Review/45664515356d4957bcb4b266acb1b9e5?revisionId=cd9e3c9c295b459882430bebab5e935d)

christothes commented on sdk/identity/Azure.Identity/src/AzureIdentityEventSource.cs:
```diff
@@ -73,6 +74,7 @@ public void GetTokenSucceeded(string method, TokenRequestContext context, DateTi
             }
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode", Justification = "Parameters to this method are primitive and are trimmer safe.")]
```
Curious - why are only these three events affected?

eerhardt replied:
The reason is because the other events all end up using the primitive overloads of WriteEvent, which aren't marked as `RequiresUnreferencedCode`. For example, there is a `WriteEvent` overload that takes 3 strings. Calling that doesn't produce a warning. But once you jump to 4 strings, there is no overload directly for that, and the compiler binds to the `params object[]` overload, which is marked as `RequiresUnreferencedCode`. Since we know that these are strings, it is safe to suppress the warning.

Note that this has been fixed when you target `net8.0` with https://github.com/dotnet/runtime/pull/83751. But since we aren't targeting `net8.0` (and 6 and 7 are still in support), we need to suppress these warnings so the user doesn't see them.

eerhardt committed:
Add expected AOT warnings for DiagnosticSource.Write

These warnings went away with https://github.com/dotnet/runtime/pull/76109, which was backported to .NET 7.0.3, but the CI build machines are still using an older 7.0.x, which don't have these warnings fixed. Adding expected warnings entries for them until we take a new .NET version in CI.

m-redding reviewed (approved):
This all looks good from my perspective.

eerhardt commented on sdk/extensions/Microsoft.Extensions.Azure/tests/compatibility/ExpectedAotWarnings.txt:
```diff
@@ -0,0 +1,11 @@
+AzureClientFactoryBuilder\.cs\(\d*\): Trim analysis warning IL2026: Microsoft\.Extensions\.Azure\.AzureClientFactoryBuilder\.\<\>c__DisplayClass\d*_\d*`2\.\<Azure\.Core\.Extensions\.IAzureClientFactoryBuilderWithConfiguration\<Microsoft\.Extensions\.Configuration\.IConfiguration\>\.RegisterClientFactory\>b__\d*\(!1,TokenCredential\): Using member 'Microsoft\.Extensions\.Azure\.ClientFactory\.Crea ...
```
The reason this is failing is because the warnings are showing up 2x in the following:

```
ILC : AOT analysis warning IL3050: Azure.Core.Pipeline.DiagnosticScope.ActivityAdapter.Dispose(): Using member 'System.Diagnostics.DiagnosticSource.Write(String,Object)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. DiagnosticSource may require creating new generic types or methods, which requires creating code at runtime. This may not work when AOT compiling. [D:\a\_work\1\s\eng\scripts\compatibility\TempAotCompatFiles\aotcompatibility.csproj] 
ILC : AOT analysis warning IL3050: Azure.Core.Pipeline.DiagnosticScope.ActivityAdapter.Dispose(): Using member 'System.Diagnostics.DiagnosticSource.Write(String,Object)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. DiagnosticSource may require creating new generic types or methods, which requires creating code at runtime. This may not work when AOT compiling. [D:\a\_work\1\s\eng\scripts\compatibility\TempAotCompatFiles\aotcompatibility.csproj] 
ILC : AOT analysis warning IL3050: Azure.Core.Pipeline.DiagnosticScope.ActivityAdapter.MarkFailed<T>(!!0,String): Using member 'System.Diagnostics.DiagnosticSource.Write(String,Object)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. DiagnosticSource may require creating new generic types or methods, which requires creating code at runtime. This may not work when AOT compiling. [D:\a\_work\1\s\eng\scripts\compatibility\TempAotCompatFiles\aotcompatibility.csproj] 
ILC : AOT analysis warning IL3050: Azure.Core.Pipeline.DiagnosticScope.ActivityAdapter.MarkFailed<T>(!!0,String): Using member 'System.Diagnostics.DiagnosticSource.Write(String,Object)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. DiagnosticSource may require creating new generic types or methods, which requires creating code at runtime. This may not work when AOT compiling. [D:\a\_work\1\s\eng\scripts\compatibility\TempAotCompatFiles\aotcompatibility.csproj] 
ILC : AOT analysis warning IL3050: Azure.Core.Pipeline.DiagnosticScope.ActivityAdapter.WriteStartEvent(): Using member 'System.Diagnostics.DiagnosticSource.Write(String,Object)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. DiagnosticSource may require creating new generic types or methods, which requires creating code at runtime. This may not work when AOT compiling. [D:\a\_work\1\s\eng\scripts\compatibility\TempAotCompatFiles\aotcompatibility.csproj] 
ILC : AOT analysis warning IL3050: Azure.Core.Pipeline.DiagnosticScope.ActivityAdapter.WriteStartEvent(): Using member 'System.Diagnostics.DiagnosticSource.Write(String,Object)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. DiagnosticSource may require creating new generic types or methods, which requires creating code at runtime. This may not work when AOT compiling. [D:\a\_work\1\s\eng\scripts\compatibility\TempAotCompatFiles\aotcompatibility.csproj] 
```

So I guess I will need to duplicate those lines in the ExpectedWarnings.txt file.

m-redding replied:
Ok that makes sense. I had opted to enforce warning count to be exact to help prevent lists from getting stale.

eerhardt committed:
Fix CI errors.

eerhardt committed:
Add CheckAOTCompat to Azure.Identity

eerhardt committed:
Fix up Identity expected aot warnings for CI.
